dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/BomTest.java

```diff
@@ -72,4 +83,64 @@ private static void assertReachable(String url) throws IOException {
       Assert.fail("Could not reach " + url + "\n" + ex.getMessage());
     }
   }
+
+  /**
+   * Asserts that the members of the {@code bom} satisfy the no-downgrade rule. This rule means that
+   * the members have the highest versions among the dependencies of them. If there's a violation,
+   * users of the BOM would see our BOM downgrading certain dependencies. Downgrading a dependency
+   * is bad practice in general because newer versions have more features (classes and methods).
+   *
+   * <p>For example, if google-http-client 1.40.1 is in the BOM, then no other libraries in the BOM
+   * must not depend on the higher version of the google-http-client.
+   *
+   * @param bom the BOM to assert with this no-downgrade rule.
+   */
+  private static void assertNoDowngradeRule(Bom bom) throws InvalidVersionSpecificationException {
+
+    boolean violationFound = false;
+    StringBuilder errorMessages = new StringBuilder();
+    Map<String, Artifact> versionlessCoordinatesToArtifact = new HashMap<>();
```

Comment:
I think `versionlessCoordinates` is too verbose; it will be clear from the code that `Artifact.makeKey(..)` does not include the version.

I think it would be more clear to refer to this map as something like `bomArtifacts` to indicate these are the artifacts declared directly by the BOM. Then you will check that transitive dependencies do not include an artifact newer than what is directly declared by the BOM.

Reply:
Updated to bomArtifacts and simplified the logic.